### DISCUSSION TITLE
Error 31 on 8" Disk Images

### DISCUSSION THREAD

**Discussion by Toast-er** [keirf/flashfloppy#410] [General]
Hi all,
yesterday I flashed a Gotek with the latest Flashfloppy F/W, v3.22.
It works fine with usual 3,5" und 5,25" formats on my MFA running CP/M 2.2.
But 2 of the 3 8" formats I would like to use alwas give me error 31.
Even the standard 8" SS/SD format, which I copied 1:1 from the CompuPro example results in error 31.
I created empty files with the corrected file size to start with, which worked for the other formats.

These are the non working formats in my IMG.CFG:

```
[::256256] # Std 8" SS/SD
cyls = 77
heads = 1
rpm = 360
gap3 = 42
interleave = 1
bps = 128     
secs = 26
mode = fm
id = 1

[::1025024] # 8" DS/DD 26Sec 256byte
cyls = 77
heads = 2
rpm = 360
bps = 256
secs = 26
mode = mfm
```

These are the working ones:

```
[::1182720] # 8" DS/DD 15sec 512byte 
cyls = 77
heads = 2
secs = 15
bps = 512
mode = mfm
rpm = 360

[::1474560] # 3,5" HD 1,44MB
cyls = 80
heads = 2
secs = 18
bps = 512
mode = mfm
rpm = 300

[::1228800] # 5,25" HD 1,2MB
cyls = 80
heads = 2
secs = 15
bps = 512
mode = mfm
rpm = 360

[::737280] # 3,5"/5,25" DD 720KB
cyls = 80
heads = 2
secs = 9
bps = 512
mode = mfm
rpm = 300
```

This is my FF.CFG:

```
interface = shugart
host = unspecified
pin02 = nc
pin34 = rdy
write-protect = no
side-select-glitch-filter = 0
track-change = instant
write-drain = instant
index-suppression = no
head-settle-ms = 12
motor-delay = ignore
chgrst = step
ejected-on-startup = no
image-on-startup = init
display-probe-ms = 0
autoselect-file-secs = 2
autoselect-folder-secs = 2
folder-sort = always
sort-priority = files
nav-mode = indexed
nav-loop = yes
twobutton-action = zero
rotary = none
display-type = auto
```

The only thing, that seems to be common to the affected formats is, that the sector size differs from 512 bytes.

Christian

**Comment by keirf** [comment]
I just copied and pasted those IMG.CFG entries into my IMG.CFG and created empty 256256-byte and 1025024-byte IMG files. They loaded up fine; no Error 31 here.

**Comment by Toast-er** [comment]
May there be something wrong in the FF.CFG file?
Is it possible something went wrong when flashing the firmware?

**Comment by Toast-er** [comment]
Still it doesn't work here.
I tried:
- another Gotek device
- another USB stick
- reprogramming the firmware 

Any hints?

**Comment by keirf** [comment]
Only place IMG.CFG and the IMG files themselves on your stick. Reset flash by holding both buttons a few seconds after power on with no stick inserted.
Possibly run the alt/logfile firmware and grab a FFLOG.TXT of attempting to mount an IMG. https://github.com/keirf/FlashFloppy/wiki/Logging-To-USB-Drive
Even if it doesn't immediately help, I can provide test firmwares with extra logging.

**Comment by keirf** [comment]
Further, basic question, but are you certain your IMG files are exactly 256256 bytes and 10251024 bytes?

**Comment by Toast-er** [comment]
OK, so here we go with only these two files: (no FF.CFG file)

```
G:\>dir
 Datenträger in Laufwerk G: ist POLLINSTICK
 Volumeseriennummer: A4E7-BD0E

 Verzeichnis von G:\

29.12.2020  17:30               648 IMG.CFG
11.11.2014  19:43           256.256 DSKA0000.IMG
               2 Datei(en),        256.904 Bytes
               0 Verzeichnis(se),  8.034.951.168 Bytes frei

G:\>
```
I used a [BASCOM](https://www.autometer.de/unix4fun/z80pack/ftp/bascom.tgz) image file.

Inserting stick -> E31
Gotek reset by pressing both buttons without stick -> rSt
Stick reinserted -> No error shown. :)
Typing some files from the disk image to screen works -> looks like the disk Image reads correct.
Test if whole disk is readable with FINDBAD.COM: No errors shown.
Trying to write to the disk image by copying a file to it with PIP -> PIP shows an error: CANNOT CLOSE DESTINATION FILE (there is enough space left in the disk image). The destination file is not show with DIR.
Re-formatting disk image with Rolf-Dieter Klein's universal formatter -> format and verfiy shows no errors. (verfy only checks if sectors are readable, not if if the ciontents is correct)
DIR shows the same filesand  typing files shows the same result as before re-formatting. -> writing to the disk image fails.

The write indicator on the 3digit 7segment led display (right digit, topmost line) flashes during write operations.

Before this test, using the FF.CFG(*) file, writing to images (1,44MB, 1,2MB, 720KB) worked.
So I think there is some setting in the FF.CFG, that refuses loading of images that have a sector size that differs from 512bytes, but there is also some setting, that is needed to get disk writing to work.

Should I run some tests with the logging firmware? If so, which tests?

Christian


(*) I used this FF.CFG file:
```
interface = shugart
side-select-glitch-filter = 50
track-change = realtime
write-drain = realtime
index-suppression = no
head-settle-ms = 10
motor-delay = ignore
chgrst = step
ejected-on-startup = no
image-on-startup = init
display-probe-ms = 0
autoselect-file-secs = 2
autoselect-folder-secs = 2
folder-sort = always
sort-priority = files
nav-mode = indexed
nav-loop = yes
twobutton-action = zero
rotary = none
display-type = auto

```

**Comment by keirf** [comment]
Try putting FF.CFG back and see what happens?

**Comment by keirf** [comment]
Hmmm ok it must have something to do with `nav-mode = indexed` I think. I can reproduce the error with your FF.CFG. So I will investigate it now.

**Comment by keirf** [comment]
Can you please try the automated build artifact linked here: https://github.com/keirf/FlashFloppy/actions/runs/452833031

The bug was that IMG.CFG is not parsed in indexed mode. Actually it was deliberate / by design but I have no idea why I did that!!

**Comment by Toast-er** [comment]
I flashed the new build. Now all defined image types read correct. :)
But writing fails on 8" ss/sd (256256 bytes) images like described above.

**Comment by keirf** [comment]
You have two different FF.CFGs listed above. They differ in settings for `write-drain` and `track-change`. Are you using the one that works/worked before (at least for some other image types)?

**Comment by keirf** [comment]
Otherwise the tiny 128-byte sectors may be slow to write back on your USB drive. I ought to batch them but I don't, currently. You may have better luck with a different USB drive. Also `write-drain` and `track-change` and `index-suppression` are worth playing with. You want them to have the default values if at all possible.

**Comment by Toast-er** [comment]
OK,  so I changed these three lines in FF.CFG:
```
# track-change = realtime
# write-drain = realtime
# index-suppression = no
```
Also, I moved to a faster USB drive (about 18MB/sec).
Unfortunately still writing so the SS/SD image does not work.
All other Image types still seem to work.

May this also be a issue with FM vs. MFM?

**Comment by keirf** [comment]
Do the default FF.CFG options work with other image types? Those would be:
```
track-change = instant
write-drain = instant
index-suppression = yes
```
Please use these options if they generally work with your host system, and see if they then work with SSSD too.

If not, it is worth gathering a FFLOG.TXT file using alt/logfile firmware, to log what happens when you try to write to a SSSD image.

**Comment by Toast-er** [comment]
With these options I get a 'delayed write error'.
changing back to
`index suppression = no `
works, at least on 1,44MB image. Trying the other ones...

**Comment by Toast-er** [comment]
still only ss/sd does not write.
will try with the logging F/W now.

**Comment by Toast-er** [comment]
I navigated to image #6, did, a write operation, and navigated to image #7 to flush the logfile.
As before, write was not successfull.
Used F/W: FF_Gotek-Logfile-f83211f-debug.upd

FFLOG.TXT:
```
** FlashFloppy v3.22 for Gotek
** Keir Fraser <keir.xen@gmail.com>
** https://github.com/keirf/FlashFloppy

Build: Dec 30 2020 14:40:06
Board: Standard
Config: Flash Slot 13 (ver 2, size 69)
Interface: Shugart (pin2=chg, pin34=rdy)
I2C: Invalid bus SCL=1 SDA=0
Display: 3-Digit LED

> USBH_USR_Init
> USBH_USR_ResetDevice
> USBH_USR_DeviceAttached
> USBH_USR_ResetDevice
> USBH_USR_DeviceSpeedDetected
> Device speed: Full
> USBH_USR_DeviceDescAvailable
 VID : 058F
 PID : 6387
> USBH_USR_DeviceAddressAssigned
> USBH_USR_ConfigurationDescAvailable
> Class connected: 08 (MSC)
 Manufacturer : Generic
 Product : Mass Storage
 Serial Number : E3798CC6
> USBH_USR_EnumerationDone
> USBH_USR_UserInput
Cache 104 items
Mode: HxC
[DSKA0000*.*]
Current slot: 0/9
Name: 'DSKA0000' Type: img
Attr: 20 Clus: 00000006 Size: 1474560
Fast Seek: 1 frags
Cache 33 items
Cache 104 items
[DSKA0009*.*]
[DSKA0008*.*]
[DSKA0007*.*]
[DSKA0006*.*]
[DSKA0006*.*]
Current slot: 6/9
Name: 'DSKA0006' Type: img
Attr: 20 Clus: 000006d2 Size: 256256
Fast Seek: 1 frags
Cache 33 items
Cache 104 items
[DSKA0007*.*]
[DSKA0007*.*]
Current slot: 7/9
Name: 'DSKA0007' Type: img
Attr: 20 Clus: 00000711 Size: 256256
```

**Comment by keirf** [comment]
No log entries for writes to the image at all. So no sectors were extracted from the write. I suppose it must be some FM thing on this system... Might have to supply you some more logging.

**Comment by keirf** [comment]
Are you doing a full disk copy or format operation, or are you simply trying to write a small file of some kind?

**Comment by keirf** [comment]
Also, does the LED display segment flash as you observed it does for writes to other image types?

**Comment by Toast-er** [comment]
The (not) logged write would have overwritten the directory (Track2) of the CP/M disk image (`ERA *.*`).
The write indicator flashed during the test.

**Comment by Toast-er** [comment]
Also a full format with verify does not produce write entries in FFLOG.TXT:

```
Fast Seek: 1 frags
Cache 33 items
> USBH_USR_DeviceDisconnected
> USBH_USR_DeInit
> USBH_USR_ResetDevice
> USBH_USR_DeviceAttached
> USBH_USR_ResetDevice
> USBH_USR_DeviceSpeedDetected
> Device speed: Full
> USBH_USR_DeviceDescAvailable
 VID : 058F
 PID : 6387
> USBH_USR_DeviceAddressAssigned
> USBH_USR_ConfigurationDescAvailable
> Class connected: 08 (MSC)
 Manufacturer : Generic
 Product : Mass Storage
 Serial Number : E3798CC6
> USBH_USR_EnumerationDone
> USBH_USR_UserInput
Cache 104 items
Mode: HxC
[DSKA0000*.*]
Current slot: 0/9
Name: 'DSKA0000' Type: img
Attr: 20 Clus: 00000006 Size: 1474560
Fast Seek: 1 frags
Cache 33 items
Cache 104 items
[DSKA0009*.*]
[DSKA0008*.*]
[DSKA0007*.*]
[DSKA0006*.*]
[DSKA0006*.*]
Current slot: 6/9
Name: 'DSKA0006' Type: img
Attr: 20 Clus: 000006d2 Size: 256256
Fast Seek: 1 frags
Cache 33 items
Cache 104 items
[DSKA0007*.*]
[DSKA0007*.*]
Current slot: 7/9
Name: 'DSKA0007' Type: img
Attr: 20 Clus: 00000711 Size: 256256
```

**Comment by keirf** [comment]
Ahhh... the problem is I have the FM data rate hardcoded to 125kbps, which is correct for micro-diskettes. Actually the correct value for 8" disks is 250kbps. For some arcane internal reason I end up with approx the right value for reads, but not for writes. I should really fix this I think...

**Comment by Toast-er** [comment]
Ony a recognized bug is a good bug. :+1:

**Comment by keirf** [comment]
You can work around this in IMG.CFG but please don't. I will have a fixed build for you to try in a few minutes.

**Comment by keirf** [comment]
Please try the following build: https://github.com/keirf/FlashFloppy/actions/runs/453116976

**Comment by Toast-er** [comment]
Until now it works fine. :+1: 
I will do some further testing.

**Comment by keirf** [comment]
Cheers. And thanks for helping to find two bugs/issues. Most support requests are user error so it makes a change ;)

**Comment by Toast-er** [comment]
Thank you very much for this great FDD emulator F/W! :+1: 

Only with detailed information you are able to fix things.
So I always try to help programmers by giving the right answers to their questions when trying to fix bugs.
Maybe it's because I very often try things that are supposed to work, but don't, just when I use software for the first time. ;)

**Comment by keirf** [comment]
I have enough little fixes now to do a new release, probably tomorrow.

**Comment by Toast-er** [comment]
That's good news. :)

**Comment by Toast-er** [comment]
There is another thing that looks odd to me:
Newly formatted CP/M disks usually contains 0xE5 in every data byte, since there is no data written after the formatting.
But a fresh reformatted flashfloppy disk image contains 0x3F in every data byte.
To be sure that this is no odd behavior of my CP/M system, I formatted a 1,44MB disk with it, and read the disk using DCF under MS-DOS. This disk, as expected, contains only 0xE5 bytes. 
CP/M dooesn't care, so one will only see this on having a closer look at it.

**Comment by keirf** [comment]
The data certainly is written by the host and not conjured up by FlashFloppy. If writes are generally working, and your IMG file originally is blank (say all zeroes), then you can be very sure that the formatting data is coming from CP/M. So the question is: Why 3F in this case? :)

**Comment by keirf** [comment]
You could also look in FFLOG.TXT after a format operation and confirm you see no weird warnings (eg CRC errors) and just normal write logging.

**Comment by Toast-er** [comment]
The logfile looks good to me.

So what is different from a normal write during formatting?

The used FDC chip should be a Mitsubishi M5W1793-02P, which should be compatible to the WD1793.
Unfortunately, the Mitshubishi datasheet only seems to contain information about the electrical interface.


FFLOG.TXT:
```
** FlashFloppy v3.22 for Gotek
** Keir Fraser <keir.xen@gmail.com>
** https://github.com/keirf/FlashFloppy

Build: Dec 30 2020 17:54:39
Board: Standard
Config: Flash Slot 15 (ver 2, size 69)
Interface: Shugart (pin2=chg, pin34=rdy)
I2C: Invalid bus SCL=1 SDA=0
Display: 3-Digit LED

> USBH_USR_Init
> USBH_USR_ResetDevice
> USBH_USR_DeviceAttached
> USBH_USR_ResetDevice
> USBH_USR_DeviceSpeedDetected
> Device speed: Full
> USBH_USR_DeviceDescAvailable
 VID : 058F
 PID : 6387
> USBH_USR_DeviceAddressAssigned
> USBH_USR_ConfigurationDescAvailable
> Class connected: 08 (MSC)
 Manufacturer : Generic
 Product : Mass Storage
 Serial Number : E3798CC6
> USBH_USR_EnumerationDone
> USBH_USR_UserInput
Cache 104 items
Mode: HxC
[DSKA0000*.*]
Current slot: 0/9
Name: 'DSKA0000' Type: img
Attr: 20 Clus: 00000006 Size: 1474560
Fast Seek: 1 frags
Cache 33 items
Cache 104 items
[DSKA0001*.*]
[DSKA0001*.*]
Current slot: 1/9
Name: 'DSKA0001' Type: img
Attr: 20 Clus: 0000016e Size: 1474560

[lost 77049]
rite 16[11]/18... 2681 us
Write 17[12]/18... 1887 us
Write 0[01]/18... 5369 us
Write 1[02]/18... 2587 us
Write 2[03]/18... 1866 us
Write 3[04]/18... 2577 us
Write 4[05]/18... 2640 us
Write 5[06]/18... 2587 us
Write 6[07]/18... 1926 us
Write 7[08]/18... 2630 us
Write 8[09]/18... 2720 us
Write 9[0a]/18... 2599 us
Write 10[0b]/18... 2573 us
Write 11[0c]/18... 2632 us
Write 12[0d]/18... 2577 us
Write 13[0e]/18... 1860 us
Write 14[0f]/18... 2606 us
Write 15[10]/18... 1890 us
Write 16[11]/18... 2608 us
Write 17[12]/18... 1931 us
Write 0[01]/18... 5360 us
Write 1[02]/18... 2662 us
Write 2[03]/18... 2607 us
Write 3[04]/18... 2595 us
Write 4[05]/18... 2587 us
Write 5[06]/18... 2629 us
Write 6[07]/18... 1832 us
Write 7[08]/18... 1835 us
Write 8[09]/18... 1889 us
Write 9[0a]/18... 2606 us
Write 10[0b]/18... 2719 us
Write 11[0c]/18... 2590 us
Write 12[0d]/18... 2613 us
Write 13[0e]/18... 2605 us
Write 14[0f]/18... 2589 us
Write 15[10]/18... 2629 us
Write 16[11]/18... 2565 us
Write 17[12]/18... 1887 us
Write 0[01]/18... 5361 us
Write 1[02]/18... 1888 us
Write 2[03]/18... 2585 us
Write 3[04]/18... 2728 us
Write 4[05]/18... 2631 us
Write 5[06]/18... 2617 us
Write 6[07]/18... 2660 us
Write 7[08]/18... 2579 us
Write 8[09]/18... 1835 us
Write 9[0a]/18... 2573 us
Write 10[0b]/18... 1870 us
Write 11[0c]/18... 2630 us
Write 12[0d]/18... 1884 us
Write 13[0e]/18... 2634 us
Write 14[0f]/18... 1919 us
Write 15[10]/18... 2646 us
Write 16[11]/18... 2585 us
Write 17[12]/18... 1885 us
Write 0[01]/18... 5375 us
Write 1[02]/18... 2636 us
Write 2[03]/18... 2620 us
Write 3[04]/18... 1883 us
Write 4[05]/18... 2593 us
Write 5[06]/18... 1871 us
Write 6[07]/18... 2603 us
Write 7[08]/18... 1883 us
Write 8[09]/18... 2635 us
Write 9[0a]/18... 2606 us
Write 10[0b]/18... 2589 us
Write 11[0c]/18... 2584 us
Write 12[0d]/18... 1843 us
Write 13[0e]/18... 1842 us
Write 14[0f]/18... 1891 us
Write 15[10]/18... 2629 us
Write 16[11]/18... 1882 us
Write 17[12]/18... 2752 us
Cache 104 items
[DSKA0000*.*]
[DSKA0000*.*]
Current slot: 0/9
Name: 'DSKA0000' Type: img
Attr: 20 Clus: 00000006 Size: 1474560
```

**Comment by Toast-er** [comment]
Another try with activated WPCom... I think that's how it looks when things go wrong:

```
Fast Seek: 1 frags
Cache 33 items
> USBH_USR_DeviceDisconnected
> USBH_USR_DeInit
> USBH_USR_ResetDevice
> USBH_USR_DeviceAttached
> USBH_USR_ResetDevice
> USBH_USR_DeviceSpeedDetected
> Device speed: Full
> USBH_USR_DeviceDescAvailable
 VID : 058F
 PID : 6387
> USBH_USR_DeviceAddressAssigned
> USBH_USR_ConfigurationDescAvailable
> Class connected: 08 (MSC)
 Manufacturer : Generic
 Product : Mass Storage
 Serial Number : E3798CC6
> USBH_USR_EnumerationDone
> USBH_USR_UserInput
Cache 104 items
Mode: HxC
[DSKA0000*.*]
Current slot: 0/9
Name: 'DSKA0000' Type: img
Attr: 20 Clus: 00000006 Size: 1474560
Fast Seek: 1 frags
Cache 33 items
IMG IDAM Bad CRC: 0150, 234
IMG IDAM Bad CRC: 6a88, 159
IMG IDAM Bad CRC: 0814, 244
IMG Bad Wr.Off: -12033
IMG DAM Unknown
IMG IDAM Bad CRC: caab, 231
Cache 104 items
[DSKA0001*.*]
[DSKA0001*.*]
Current slot: 1/9
Name: 'DSKA0001' Type: img
Attr: 20 Clus: 0000016e Size: 1474560
```

**Comment by keirf** [comment]
I think it depends on the FDC but there is either a Format Track or Write Track command. Either way the host has control over the sector data, it's not prescribed by the FDC.

**Comment by keirf** [comment]
WPCom = write precompensation? Doesn't seem to be liked :) Of course FlashFloppy does not need any precomp.

**Comment by Toast-er** [comment]
Of course not. :)
Seem to be bad news for systems which can not deactivate write precompensation.
I think this could happen to some old machines with 8" drives.

**Comment by Toast-er** [comment]
Made another try with a Siemens SAB1793-02P FDC Chip.
Behaves the same as the Mitsubishi chip.

**Comment by keirf** [comment]
I may have to explicitly handle precomp in that case. It is used quite generally including on 3.5/5.25 micro-diskette systems and I've had no problems before. Either the precomp here is very extreme, or we've still got the bitrate a bit wrong, and precomp puts us "over the edge" for correct decoding.

**Comment by keirf** [comment]
Is it feasible for you to get scope or logic analyser traces of WDATA when precomp is enabled? On a scope just a few bit times could suffice.

**Comment by keirf** [comment]
By the way -- although FDC controls when precomp is applied, on these earlier systems the actual time shifting is not integrated into the FDC. So it is no surprise that switching the FDC chip had no effect.

**Comment by Toast-er** [comment]
I could use LA and scope, but I don't think this is worth it.
We only added the WPCom circuit to the FDC to try things out, and we found that we don't need it.
And I think you are right, it may be extreme.
All 3,5", 5,25" and 8" formats are working without WPCom. 
Only 3,5" ED (2,88MB) is missing, but this is another project for another FDC chip (intel 82077SL). ;)

**Comment by keirf** [comment]
Issues now fixed in official release v3.23
